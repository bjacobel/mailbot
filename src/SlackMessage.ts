import { ChatPostMessageArguments /* WebClient */ } from "@slack/web-api";
import { EmailAddress, HeaderValue } from "mailparser";

export type Headers = Map<string, string | HeaderValue>;

export default class SlackMessage {
  private body: string;
  private headers: Headers;
  // private slack: WebClient;

  constructor(emailBody: string, emailHeaders: Headers) {
    this.body = emailBody;
    this.headers = emailHeaders;
    // this.slack = new WebClient();
  }

  public async composeMessage(): Promise<ChatPostMessageArguments> {
    return {
      mrkdwn: true,
      username: this.fullname(),
      channel: "",
      text: this.body, // todo: transform html to markdown
    };
  }

  public async send(): Promise<void> {
    return Promise.resolve();
    // const args = await this.composeMessage();
    // log.info(JSON.stringify(args));
    // this.slack.chat.postMessage({
    //   ...args,
    // });
  }

  public fullname(): string {
    if (!this.headers.has("from")) {
      return "Unknown sender";
    }

    const fromHeader = this.headers.get("from");
    if (typeof fromHeader === "string") {
      return fromHeader.split("<")[0].trim();
    }

    if (typeof fromHeader === "object" && "value" in fromHeader) {
      return (fromHeader.value[0] as EmailAddress).name;
    }

    return "Unknown sender";
  }

  public async avatar(): Promise<string> {
    return Promise.resolve("");
  }
}
