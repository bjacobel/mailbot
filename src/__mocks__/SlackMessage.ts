// tslint:disable no-empty
const SlackMessage = jest.fn();

SlackMessage.prototype._transform = jest.fn(() => {});

export default SlackMessage;
