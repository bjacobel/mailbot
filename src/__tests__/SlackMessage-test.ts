import SlackMessage from "../SlackMessage";

jest.mock("../constants", () => ({
  ...jest.requireActual("../constants"),
  IS_PROD: false,
}));

describe("parse function handler", () => {
  describe("basic info", () => {
    it("pulls the fulname from headers when `from` is set", () => {
      expect(
        new SlackMessage(
          "",
          new Map([["from", "Brian Jacobel <bjacobel@gmail.com>"]]),
        ).fullname(),
      ).toEqual("Brian Jacobel");
    });

    it("defaults to a username when from is not set", () => {
      expect(new SlackMessage("", new Map()).fullname()).toEqual(
        "UVDSA Announcements",
      );
    });
  });

  describe("composeMessage method", () => {
    it("matches snapshot for a basic email", async () => {
      const email = "";
      const headers = new Map([["from", "Brian Jacobel <bjacobel@gmail.com>"]]);
      const slack = new SlackMessage(email, headers);
      expect(await slack.composeMessage()).toMatchSnapshot();
    });
  });
});
