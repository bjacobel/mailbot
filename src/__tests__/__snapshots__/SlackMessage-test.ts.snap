// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse function handler composeMessage method matches snapshot for a basic email 1`] = `
Object {
  "channel": "",
  "mrkdwn": true,
  "text": "",
  "username": "Brian Jacobel",
}
`;
